presto-native-execution/presto_cpp/main/runtime-metrics/prometheus-reporter/PrometheusReporter.cpp

```diff
@@ -0,0 +1,238 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "presto_cpp/main/runtime-metrics/prometheus-reporter/PrometheusReporter.h"
+
+#include <prometheus/collectable.h>
+#include <prometheus/counter.h>
+#include <prometheus/gauge.h>
+#include <prometheus/histogram.h>
+#include <prometheus/registry.h>
+#include <prometheus/summary.h>
+#include <prometheus/text_serializer.h>
+
+namespace facebook::presto::prometheus {
+
+using namespace ::prometheus;
+
+static constexpr std::string_view kSummarySuffix("_summary");
+std::unordered_map<velox::StatType, std::string> kStatTypeToStr{
+    {velox::StatType::COUNT, "Count"},
+    {velox::StatType::SUM, "Sum"},
+    {velox::StatType::AVG, "Avg"},
+    {velox::StatType::RATE, "Rate"},
+};
+struct PrometheusReporter::Impl {
+  Impl(const Labels& labels) {
+    registry = std::make_shared<::prometheus::Registry>();
+    for (auto& itr : labels) {
+      this->labels[itr.first] = itr.second;
+    }
+  }
+
+ public:
+  ::prometheus::Labels labels;
+  std::shared_ptr<::prometheus::Registry> registry;
+};
+
+PrometheusReporter::PrometheusReporter(
+    const std::map<std::string, std::string>& labels) {
+  impl_ = std::make_shared<Impl>(labels);
+}
+
+void PrometheusReporter::registerMetricExportType(
+    const char* key,
+    facebook::velox::StatType statType) const {
+  if (registeredMetricsMap_.count(key)) {
+    // Already registered;
+    VLOG(1) << "Trying to register already registered metric " << key;
+    return;
+  }
+  // '.' is replaced with '_'.
+  std::string sanitizedMetricKey = std::string(key);
+  std::replace(sanitizedMetricKey.begin(), sanitizedMetricKey.end(), '.', '_');
+  switch (statType) {
+    case facebook::velox::StatType::COUNT: {
+      // A new MetricFamily object is built for every new metric key.
+      auto& counterFamily =
+          BuildCounter().Name(sanitizedMetricKey).Register(*impl_->registry);
+      auto& counter = counterFamily.Add(impl_->labels);
+      registeredMetricsMap_.emplace(
+          std::string(key), StatsInfo{statType, &counter});
+      break;
+    }
+    case facebook::velox::StatType::SUM:
+    case facebook::velox::StatType::AVG:
+    case facebook::velox::StatType::RATE: {
+      auto& gaugeFamily =
+          BuildGauge().Name(sanitizedMetricKey).Register(*impl_->registry);
+      auto& gauge = gaugeFamily.Add(impl_->labels);
+      registeredMetricsMap_.emplace(
+          std::string(key), StatsInfo{statType, &gauge});
+      break;
+    }
+    default:
+      VELOX_UNSUPPORTED(
+          "Unsupported metric type {}", static_cast<int>(statType));
```

Comment:
Use `kStatTypeToStr` here?

Reply:
I added statTypeString on the Velox side as well https://github.com/facebookincubator/velox/pull/10099

Reply:
Creating HISTOGRAM type is not a blocker, just for completeness we need to add it. The APIs for registration and adding metric values are separated and we don't really use the HISTOGRAM type as of now.

Reply:
We should consider adding a different map for histogram or use this enum type. We must avoid workarounds such as using COUNT as a placeholder.